design/one-pager-generic-usage-type.md

```diff
@@ -0,0 +1,397 @@
+
+## Proposal
+
+We propose introducing a new `Usage` type that can be used to declare usage
+relationships between Crossplane resources. This type will be defined by the
+Core Crossplane repository, and be available for use as part of a composition.
+
+The relations defined by this type will be enforced by an admission webhook that
+will be running as part of the Crossplane core. Similar to the upstreams [liens
+proposal], the webhook will reject deletions of resources that are in use by
+other resources with a ["409 Conflict" error code].
+
+### API
+
+We will introduce a new `Usage` type as follows:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: Usage
+metadata:
+  name: release-uses-cluster
+spec:
+  of:
+  - apiVersion: eks.upbound.io/v1beta1
+    kind: Cluster
+    name: my-cluster
+  by:
+    apiVersion: helm.crossplane.io/v1beta1
+    kind: Release
+    name: my-prometheus-chart
+```
+
+The `spec.by` field will define the resource that will be using the resources
+defined in `spec.of`. Both will support only cluster-scoped resources, namely
+`Composites` and `Managed Resources`.
+
+In this example, we define a usage relationship between a Helm `Release` and an
+EKS `Cluster`. The deletion of the `Cluster` resource will be rejected as long
+as this `Release` resource exists.
+
+In addition to direct referencing with names, we will also support selectors to
+match labels or controller reference. This will allow us to define usage
+relationships between resources created by the same composition instance.
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: Usage
+metadata:
+  name: release-uses-cluster
+spec:
+  of:
+  - apiVersion: eks.upbound.io/v1beta1
+    kind: Cluster
+    selector:
+      matchControllerRef: true
+  by:
+    apiVersion: helm.crossplane.io/v1beta1
+    kind: Release
+    selector:
+      matchControllerRef: true
+```
+
+While the well-known use cases for this feature are solved with a one-to-one
+usage relationship, we will support one-to-many relationship 
+(i.e. `spec.of` as a list) as well, considering it provides more flexibility
+without introducing additional complexity to the implementation.
+
+### Implementation
+
+The implementation of this feature will be done in two parts:
+
+1. The `Usage` type and its controller, which will manage the lifecycle of the
+   resource.
+2. The admission webhook that will enforce the usage relationships.
+
+Both will be implemented in the Crossplane core repository and would not require
+any changes in the providers.
+
+#### Usage Controller
+
+The usage controller will be responsible for managing the lifecycle of the usage
+relationship by:
+
+- Resolving selectors, if any.
+- Adding owner reference from the using resource (i.e., `spec.by`) to the `Usage`.
+- Preventing deletion of `Usage` before the using resource is deleted.
+- Adding or removing admission webhook rules to intercept the `DELETE` requests
```

Comment:
Rules do not include the name of the resource so we would need one rule per GVR, so I don't think we would need to scale up to thousands of rules. If every installed type has at least one dependency (which is not likely), the number of rules would equal to the number of CRDs + the number of XR types. I don't think this would be a problem in practice.

However, using a selector sounds like a good idea if we observe this as a problem. I am wondering if we could instead add / remove a well defined label on the used resource by the controller and have a fixed webhook rule for all objects but matching that label. This way, the controller wouldn't need to interact with the webhook rules but would need to add/remove labels on the used resource. This could even be helpful to reset a client's backoff for deletion, e.g. garbage collector.

Something like:

```
webhooks:
- name: nousages.apiextensions.crossplane.io
  objectSelector:
    matchLabels:
      crossplane.io/in-use: true
  rules:
  - operations: ["DELETE"]
    apiGroups: ["*"]
    apiVersions: ["*"]
    resources: ["*"]
    scope: "*"
```

Reply:
I've updated the design with the label/selector approach above.